src/util/interval.cpp

```diff
@@ -0,0 +1,1975 @@
+  }
+
+  return get_extremes(*this, o, ID_shl);
+}
+
+// Arithmetic
+constant_interval_exprt
+constant_interval_exprt::right_shift(const constant_interval_exprt &o) const
+{
+  if(o.is_single_value_interval() && is_single_value_interval())
+  {
+    return handle_constant_binary_expression(o, ID_ashr);
+  }
+
+  if(is_negative(o.get_lower()))
+  {
+    return top();
+  }
+
+  return get_extremes(*this, o, ID_ashr);
+}
+
+constant_interval_exprt
+constant_interval_exprt::bitwise_xor(const constant_interval_exprt &o) const
+{
+  if(o.is_single_value_interval() && is_single_value_interval())
+  {
+    return handle_constant_binary_expression(o, ID_bitxor);
+  }
+
+  return top();
+}
+
+constant_interval_exprt
+constant_interval_exprt::bitwise_or(const constant_interval_exprt &o) const
+{
+  if(o.is_single_value_interval() && is_single_value_interval())
+  {
+    return handle_constant_binary_expression(o, ID_bitor);
+  }
+
+  return top();
+}
+
+constant_interval_exprt
+constant_interval_exprt::bitwise_and(const constant_interval_exprt &o) const
+{
+  if(o.is_single_value_interval() && is_single_value_interval())
+  {
+    return handle_constant_binary_expression(o, ID_bitand);
+  }
+
+  return top();
+}
+
+constant_interval_exprt constant_interval_exprt::bitwise_not() const
+{
+  if(is_single_value_interval())
+  {
+    return handle_constant_unary_expression(ID_bitnot);
+  }
+
+  return top();
+}
+
+tvt constant_interval_exprt::less_than(const constant_interval_exprt &o) const
+{
+  // [get_lower, get_upper] < [o.get_lower(), o.get_upper()]
+  if(is_single_value_interval() && o.is_single_value_interval())
+  {
+    return tvt(less_than(get_lower(), o.get_lower()));
+  }
+
+  if(less_than(get_upper(), o.get_lower()))
+  {
+    return tvt(true);
+  }
+
+  if(greater_than(get_lower(), o.get_upper()))
+  {
+    return tvt(false);
+  }
+
+  return tvt::unknown();
+}
+
+tvt constant_interval_exprt::greater_than(
+  const constant_interval_exprt &o) const
+{
+  return o.less_than(*this);
+}
+
+tvt constant_interval_exprt::less_than_or_equal(
+  const constant_interval_exprt &o) const
+{
+  if(is_single_value_interval() && o.is_single_value_interval())
+  {
+    return tvt(less_than_or_equal(get_lower(), o.get_lower()));
+  }
+
+  // [get_lower, get_upper] <= [o.get_lower(), o.get_upper()]
+  if(less_than_or_equal(get_upper(), o.get_lower()))
+  {
+    return tvt(true);
+  }
+
+  if(greater_than(get_lower(), o.get_upper()))
+  {
+    return tvt(false);
+  }
+
+  return tvt::unknown();
+}
+
+tvt constant_interval_exprt::greater_than_or_equal(
+  const constant_interval_exprt &o) const
+{
+  return o.less_than_or_equal(*this);
+}
+
+tvt constant_interval_exprt::equal(const constant_interval_exprt &o) const
+{
+  if(is_single_value_interval() && o.is_single_value_interval())
+  {
+    return tvt(equal(get_lower(), o.get_lower()));
+  }
+
+  if(equal(get_upper(), o.get_upper()) && equal(get_upper(), o.get_upper()))
+  {
+    return tvt(true);
+  }
```

Comment:
This seems wrong. For one, it appears to be `if(A && A)` and if the intention is whether upper == o.upper && lower ==o.lower, this is wrong. I think this should always be unknown unless the intersection is empty, then false, or one element, then true

Reply:
TBH the semantics of this are confusing, do you expect [a, b] == [a, b] to be true or unknown The docs appear to suggest unknown, this behaviour appears to suggest true. I think depending on context you will want one or the other, so probably much safer to not guess and implement two clearly different methods: `range_equal` which returns true or false if the ranges are exactly the same, and value_equal which returns a tvt - but this can be left for now and fixed in the follow up PR